File: src/main/java/com/xpdustry/nucleus/database/PostgresDatabase.java
// SPDX-License-Identifier: GPL-3.0-only
package com.xpdustry.nucleus.database;

import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.time.Instant;
import java.util.List;
import java.util.Optional;
import org.intellij.lang.annotations.Language;

public interface PostgresDatabase {

    <R> R withTransaction(final SQLFunction<Transaction, R> function);

    Connection newOrphanConnection() throws SQLException;

    interface Transaction {
        StatementBuilder prepareStatement(final @Language("SQL") String statement) throws SQLException;
    }

    interface StatementValueBinder<B extends StatementValueBinder<B>> {

        B bind(final String value) throws SQLException;

        B bind(final int value) throws SQLException;

        B bind(final long value) throws SQLException;

        B bind(final boolean value) throws SQLException;

        B bind(final float value) throws SQLException;

        B bind(final double value) throws SQLException;

        B bind(final byte[] value) throws SQLException;

        B bind(final Instant value) throws SQLException;
    }

    interface StatementBuilder extends StatementValueBinder<StatementBuilderExecute>, StatementBuilderExecute {

        <T> StatementBuilderExecute batch(
                final Iterable<T> values, final SQLBiConsumer<StatementValueBinder<?>, T> consumer) throws SQLException;
    }

    interface StatementBuilderExecute extends StatementValueBinder<StatementBuilderExecute> {

        <T> List<T> executeSelect(final SQLFunction<ResultSet, T> mapper) throws SQLException;

        <T> Optional<T> executeSingleSelect(final SQLFunction<ResultSet, T> mapper) throws SQLException;

        int executeUpdate() throws SQLException;

        boolean executeSingleUpdate() throws SQLException;
    }
}

File: src/main/kotlin/com/xpdustry/nucleus/database/PostgresDatabase.kt
// SPDX-License-Identifier: GPL-3.0-only
package com.xpdustry.nucleus.database

import kotlin.ByteArray
import kotlin.Float
import kotlin.Int
import kotlin.Long
import kotlin.String
import kotlin.time.Instant
import kotlinx.coroutines.Job
import org.intellij.lang.annotations.Language

interface PostgresDatabase {
    suspend fun <R> transaction(block: suspend Handle.() -> R): R

    fun listen(channel: String, onNotification: suspend (String) -> Unit): Job

    interface Handle {
        fun @receiver:Language("SQL") String.asPreparedStatement(): PreparedStatementBuilder
    }

    interface PreparedStatementBuilder {
        fun push(value: String): PreparedStatementBuilder

        fun push(value: Int): PreparedStatementBuilder

        fun push(value: Long): PreparedStatementBuilder

        fun push(value: Boolean): PreparedStatementBuilder

        fun push(value: Double): PreparedStatementBuilder

        fun push(value: Float): PreparedStatementBuilder

        fun push(value: ByteArray): PreparedStatementBuilder

        fun push(value: Instant): PreparedStatementBuilder

        suspend fun <T> executeSelect(mapper: Row.() -> T): List<T>

        suspend fun <T> executeSingleSelect(mapper: Row.() -> T): T

        suspend fun executeUpdate(): Int

        suspend fun executeSingleUpdate(): Boolean
    }

    interface Row {
        fun getString(name: String): String?

        fun getInt(name: String): Int?

        fun getLong(name: String): Long?

        fun getBoolean(name: String): Boolean?

        fun getDouble(name: String): Double?

        fun getFloat(name: String): Float?

        fun getBytes(name: String): ByteArray?

        fun getInstant(name: String): Instant?
    }
}
